plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/procedure/RegisterTableProcedure.java

```diff
@@ -0,0 +1,226 @@
+        }
+        catch (ReflectiveOperationException e) {
+            throw new AssertionError(e);
+        }
+    }
+
+    private final TrinoCatalogFactory catalogFactory;
+    private final TrinoFileSystemFactory fileSystemFactory;
+    private final ClassLoader classLoader;
+
+    @Inject
+    public RegisterTableProcedure(TrinoCatalogFactory catalogFactory, TrinoFileSystemFactory fileSystemFactory)
+    {
+        this.catalogFactory = requireNonNull(catalogFactory, "catalogFactory is null");
+        this.fileSystemFactory = requireNonNull(fileSystemFactory, "fileSystemFactory is null");
+        // this class is loaded by PluginClassLoader, and we need its reference to be stored
+        this.classLoader = getClass().getClassLoader();
+    }
+
+    @Override
+    public Procedure get()
+    {
+        return new Procedure(
+                SYSTEM_SCHEMA,
+                PROCEDURE_NAME,
+                ImmutableList.of(
+                        new Procedure.Argument(SCHEMA_NAME, VARCHAR),
+                        new Procedure.Argument(TABLE_NAME, VARCHAR),
+                        new Procedure.Argument(TABLE_LOCATION, VARCHAR),
+                        new Procedure.Argument(METADATA_FILE_NAME, VARCHAR, false, null)),
+                REGISTER_TABLE.bindTo(this));
+    }
+
+    public void registerTable(
+            ConnectorSession clientSession,
+            ConnectorAccessControl accessControl,
+            String schemaName,
+            String tableName,
+            String tableLocation,
+            String metadataFileName)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
+            doRegisterTable(
+                    clientSession,
+                    accessControl,
+                    schemaName,
+                    tableName,
+                    tableLocation,
+                    Optional.ofNullable(metadataFileName));
+        }
+    }
+
+    private void doRegisterTable(
+            ConnectorSession clientSession,
+            ConnectorAccessControl accessControl,
+            String schemaName,
+            String tableName,
+            String tableLocation,
+            Optional<String> metadataFileName)
+    {
+        checkProcedureArgument(schemaName != null, "schema_name cannot be null");
+        checkProcedureArgument(tableName != null, "table_name cannot be null");
+        checkProcedureArgument(tableLocation != null, "table_location cannot be null");
+        if (metadataFileName.isPresent()) {
+            validateMetadataFileName(metadataFileName.get());
+        }
+        SchemaTableName schemaTableName = new SchemaTableName(schemaName, tableName);
+        accessControl.checkCanCreateTable(null, schemaTableName, ImmutableMap.of());
```

Comment:
Is the check sufficient for authorization check?
How to test authorization from tests? Any reference of existing test cases which checks for authorization?

CC: @findepi | @kokosing

Reply:
> Is the check sufficient for authorization check?

I am not sure what is the semantic of this procedure. If it is equivalent to create table then yes.

See the example of test of access control for procedure: `io.trino.plugin.hive.BaseHiveConnectorTest#testCreateEmptyNonBucketedPartition`

Reply:
technical: we cannot pass `null` for `ConnectorSecurityContext`; it looks like a hack and can fail at runtime

> If it is equivalent to create table then yes.

it's equivalent to "create external table" in Hive -- new table object, with existing data from storage

for "create external table" we have the `hive.non-managed-table-creates-enabled`
for `register_partition` we also have `hive.allow-register-partition-procedure` toggle

cc @alexjo2144 @ilfrin since we discussed this recently
cc @electrum

Reply:
Added `iceberg.allow-register-table-procedure` flag to toggle the `register_table` feature. By default it is set to `false`